armbotoffboard/subsystems/drivesubsystem.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) FIRST and other WPILib contributors.
+# Open Source Software; you can modify and/or share it under the terms of
+# the WPILib BSD license file in the root directory of this project.
+
+import commands2
+import constants
+import wpilib
+import wpilib.drive
+import typing
+
+
+class DriveSubsystem(commands2.SubsystemBase):
+    def __init__(self) -> None:
+        super().__init__()
+
+        # The motors on the left side of the drive.
+        self.left = wpilib.MotorControllerGroup(
+            wpilib.PWMSparkMax(constants.kLeftMotor1Port),
+            wpilib.PWMSparkMax(constants.kLeftMotor2Port),
+        )
+
+        # The motors on the right side of the drive.
+        self.right = wpilib.MotorControllerGroup(
+            wpilib.PWMSparkMax(constants.kRightMotor1Port),
+            wpilib.PWMSparkMax(constants.kRightMotor2Port),
+        )
+
+        # The robot's drive
+        self.drive = wpilib.drive.DifferentialDrive(self.left, self.right)
+
+        # The left-side drive encoder
+        self.left_encoder = wpilib.Encoder(
+            constants.kLeftEncoderPorts[0],
+            constants.kLeftEncoderPorts[1],
+            constants.kLeftEncoderReversed,
+        )
+
+        # The right-side drive encoder
+        self.right_encoder = wpilib.Encoder(
+            constants.kRightEncoderPorts[0],
+            constants.kRightEncoderPorts[1],
+            constants.kRightEncoderReversed,
+        )
+
+        # Creates a new drive subsystem
+        self.left_encoder.setDistancePerPulse(constants.kEncoderDistancePerPulse)
+        self.right_encoder.setDistancePerPulse(constants.kEncoderDistancePerPulse)
+
+        # We need to invert one side of the drivetrain so that positive voltages
+        # result in both sides moving forward. Depending on how your robot's
+        # gearbox is constructed, you might have to invert the left side instead.
+        self.right.setInverted(True)
+
+    def arcadeDrive(
+        self, fwd: typing.Callable[[], float], rot: typing.Callable[[], float]
+    ) -> None:
+        """Drives the robot using arcade controls.
+
+        Args:
+            fwd: The commanded forward movement
+            rot: The commanded rotation
+        """
+        self.drive.arcadeDrive(fwd(), rot(), True)
```

Comment:
Passing a `Callable` in here seems rather odd; it looks like this method is used in a lambda already making this even weirder. Is this really what the Java example does?

Reply:
I follow the logic of a [command](https://github.com/robotpy/examples/blob/main/commands-v2/hatchbot/commands/defaultdrive.py). The example executes the `arcadeDrive` directly from the subsystem.
Well, that was the way I understood to fix this issue.

Reply:
Yep, and that example does make sense, as the callables are kept around and so the commanded speeds can change over time. This method was immediately throwing away the callables, so adding an unused layer of indirection.